### PR TITLE
feat(recipes): add VLM knowledge distillation recipe with chunked KD loss

### LINKED ISSUE #2195: Knowledge distillation recipe for VLMs
Hi NeMo AutoModel team,

Thanks again for the great work on the project. I noticed a small but useful gap on the VLM side that I would like to help close.

## Background

`recipes/llm/kd.py` already provides a complete LLM knowledge-distillation recipe (`(1 - kd_ratio) * ce_loss + kd_ratio * kl_loss(student || teacher)`, frozen teacher, PP-aware teacher logit injection, FSDP2 / TP support). The corresponding piece for VLMs does not exist yet — `recipes/vlm/` only contains `finetune.py`.

This is a natural fit for the SFT-only positioning of AutoModel: KD uses a frozen teacher with fully supervised label + logit alignment, no reward model and no policy optimization, so it stays inside the framework's scope and does not overlap with NeMo-RL.

## Motivation

Distilling large VLMs into smaller deployable ones is a common production need:

- **Qwen3-VL-72B → Qwen3-VL-7B / 4B**
- **DeepSeek-VL → smaller in-house VLM**
- **Mistral-Medium-3.5 (128B FP8) → 12B**

Today users have to fork `recipes/llm/kd.py` and reapply most of the multimodal handling already living in `recipes/vlm/finetune.py`. A first-class VLM KD recipe would let them stay inside the framework and benefit from the existing parallelism and checkpointing work.

## Proposed implementation

I would like to take this on. The plan mirrors the LLM KD recipe and reuses the existing VLM training loop:

1. **New recipe**: `nemo_automodel/recipes/vlm/kd.py`, structured like the LLM version:
   - Frozen teacher built via `NeMoAutoModelForImageTextToText.from_pretrained(...)` (no PEFT / no FP8 / no QAT for the teacher).
   - Student built via the existing `recipes/vlm/finetune.py` builder path so all current PEFT / freeze / quantization / parallelism options continue to work.
   - Loss: reuse `components/loss/kd_loss.py`; total loss = `(1 - kd_ratio) * masked_ce + kd_ratio * kl_div(student, teacher)`.

2. **VLM-specific handling** (the part that is not copy-paste from the LLM recipe):
   - **Processor over tokenizer**: feed both teacher and student the same `pixel_values` / `image_grid_thw` / `attention_mask` produced by the existing VLM collate functions, instead of separate tokenizer calls.
   - **Image-token positions in the KL term**: image placeholder positions get replaced by vision features in both models. By default, mask the KL contribution at those positions (their distributions can be very narrow and hurt KD signal); make it configurable via `kd.kl_on_image_tokens: bool = False` so users can flip it for ablation.
   - **PP + chunk-aware teacher**: the teacher must work alongside the recently introduced `_pp_keep_self_forward` flag (PR #2192) so chunk-aware VLMs (Qwen3-VL-MoE / KimiVL / Kimi-K2.5-VL / Qwen3.5-MoE) keep their own forward when used as a teacher. The student path already has this; the teacher build needs to skip `patch_hf_model_for_pp` for these classes too.
   - **`freeze_vision_tower=true` on the student** should be respected and is the recommended default for KD (the student usually does not need vision-tower updates when copying a strong teacher).
   - **Vocabulary alignment check**: assert teacher and student share `tokenizer.vocab_size` and the same image / video / audio placeholder ids, since the KL term assumes per-token distribution alignment. Raise a clear error otherwise rather than silently producing wrong gradients.

3. **Example configs** under `examples/vlm_finetune/kd/`, e.g. `qwen3_vl_72b_to_7b_medpix.yaml` and `qwen3_vl_moe_235b_to_30b_neat_packing.yaml`.

4. **Tests**:
   - Unit test for the recipe's loss (CE + KL combination, matches the LLM KD test pattern).
   - Functional test on a tiny VLM teacher / student pair with a real `pixel_values` batch, verifying that loss decreases over a few steps.
   - Regression test for the image-token KL mask.

## Questions for the maintainers before I start

1. Is this direction welcome as a community contribution? (No objections so far on RL-vs-SFT scope, since KD is supervised.)
2. Preferred file location: `recipes/vlm/kd.py` mirroring the LLM layout, or somewhere else?
3. Any preference for image-token KL behavior — mask by default vs. include by default?
4. Are there existing internal benchmarks / known-good distillation pairs that the example configs should target so the recipe lands with credible numbers?

I am happy to follow whatever conventions you prefer and to keep the scope tight to one recipe + tests + a couple of example configs. Looking forward to your guidance.

### PR DESCRIPTION
**Pull request by khazic**
# What does this PR do ?

Adds a knowledge distillation recipe for vision-language models, plus a memory-efficient chunked path in the existing KD loss to keep peak fp32 memory bounded for large-vocab VLMs.

Closes #2195.

# Changelog

- `feat(loss)`: add `chunk_size` parameter to `KDLoss` that materializes the fp32 `[chunk_size, vocab_size]` probability matrix one chunk at a time. Numerically identical to the unchunked path; TP path unchanged. New unit tests cover chunked vs. unchunked equivalence, temperature scaling, and `num_batch_labels`.
- `feat(recipes)`: add `KnowledgeDistillationRecipeForVLM` under `nemo_automodel/recipes/vlm/kd.py`. Mirrors the LLM KD recipe structure: frozen teacher `NeMoAutoModelForImageTextToText`, KD loss term, `kd_ratio` linear mix between CE and KD. Forwards multimodal inputs (`pixel_values`, `image_grid_thw`, etc.) to both teacher and student; eagerly frees intermediate activations; reports CE/KD sub-losses in validation metrics. Respects `freeze_vision_tower` on the student. Pipeline parallelism is not supported in this initial version.
- `docs(examples)`: add `examples/vlm_kd/qwen3_5/qwen3_5_vl_4b_kd.yaml` distilling Qwen3.5-9B → Qwen3.5-4B on the public `mmoukouba/MedPix-VQA` dataset. Exercises chunked KD (`chunk_size: 512`), freezes vision and audio towers on the student, FSDP2.

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Automodel/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?

# Additional Information

- Related to #2195 (Knowledge distillation recipe for VLMs).
- Unit tests: `pytest tests/unit_tests/loss/test_kd_loss.py` — 28 passed locally.

### CODE REVIEW

**Comment by khazic** [comment]
Full 300-step sanity run on the example config — Qwen3.5-9B (teacher) → Qwen3.5-4B (student) on `mmoukouba/MedPix-VQA`, FSDP2 on 8×H100, `chunk_size=512`, `kd_ratio=0.5`, `temperature=1.0`. Total wall time ≈ 71 min.

Per-step totals match `(1-kd_ratio)*ce + kd_ratio*kd` exactly (e.g. step 75: `0.5*1.7002 + 0.5*2.1826 = 1.9414` ✓). No NaN, no divergence; eager-free of intermediate activations keeps peak memory bounded across image-size variance.

| window | steps | ce mean | kd mean | total mean | peak mem |
|---|---|---|---|---|---|
| early | 75–99   | 1.69 | 2.03 | 1.86 | 33.5 GiB |
| mid   | 126–149 | 1.62 | 1.99 | 1.80 | 45.6 GiB |
| final | 276–299 | **1.29** | **1.92** | **1.60** | 36.0 GiB |

CE loss drops ~24% (1.69 → 1.29) end-to-end; KD loss drops ~5% (2.03 → 1.92). The asymmetry is expected — forward KL = `H(P_teacher) + KL(P_teacher‖P_student)`, and the teacher-entropy term is a constant offset that does not decay with training, so the reducible signal is concentrated in the CE component.

<details><summary>Raw log excerpt (early / mid / final windows)</summary>

```
step  75 | loss 1.9414 | ce 1.7002 | kd 2.1826 | lr 4.65e-05 | mem 27.6 GiB | tps 1295
step  80 | loss 1.6199 | ce 1.5234 | kd 1.7163 | lr 4.58e-05 | mem 33.5 GiB | tps 1176
step  90 | loss 1.7475 | ce 1.4456 | kd 2.0494 | lr 4.40e-05 | mem 28.5 GiB | tps  956
step  99 | loss 1.9515 | ce 1.7086 | kd 2.1944 | lr 4.22e-05 | mem 27.7 GiB | tps 1078
step 126 | loss 1.7391 | ce 1.4892 | kd 1.9889 | lr 3.58e-05 | mem 27.4 GiB | tps  962
step 131 | loss 1.9853 | ce 1.6907 | kd 2.2799 | lr 3.45e-05 | mem 45.6 GiB | tps 1182
step 140 | loss 1.4328 | ce 1.1651 | kd 1.7005 | lr 3.21e-05 | mem 36.9 GiB | tps  953
step 149 | loss 1.7113 | ce 1.5619 | kd 1.8607 | lr 2.95e-05 | mem 32.3 GiB | tps 1040
step 280 | loss 1.4842 | ce 1.0968 | kd 1.8717 | lr 1.10e-06 | mem 26.5 GiB | tps 1182
step 285 | loss 1.4599 | ce 1.0999 | kd 1.8199 | lr 8.28e-07 | mem 27.4 GiB | tps  955
step 294 | loss 1.4891 | ce 1.1008 | kd 1.8775 | lr 5.42e-07 | mem 30.9 GiB | tps  919
step 298 | loss 1.3651 | ce 1.0605 | kd 1.6697 | lr 5.02e-07 | mem 30.7 GiB | tps 1120
step 299 | loss 1.4761 | ce 1.3185 | kd 1.6337 | lr 5.00e-07 | mem 33.6 GiB | tps  996
```

</details>

**Comment by khazic** [comment]
Clarification on scope: this PR intentionally implements the initial/base VLM KD path only, targeting dense models with the current example configuration (`tp_size=1`, `ep_size=1`, `pp_size=1`).

Support and validation for tensor parallelism, expert parallelism, pipeline parallelism, and MoE-specific behavior will be handled in a follow-up PR. This PR is meant to establish the basic dense VLM KD recipe and chunked KD loss path first.

**Review by HuiyingLi — APPROVED** [review]
Thank you! Looking forward to followup prs.
